pkg/util/ip/ip.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ip
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"sort"
+
+	"github.com/vmware-tanzu/antrea/pkg/apis/networking/v1beta1"
+)
+
+const (
+	v4BitLen = 8 * net.IPv4len
+	v6BitLen = 8 * net.IPv6len
+)
+
+// This function takes in one allow CIDR and multiple except CIDRs and gives diff CIDRs
+// in allowCIDR eliminating except CIDRs. It currently supports only IPv4. except CIDR input
+// can be changed.
+func DiffFromCIDRs(allowCIDR *net.IPNet, exceptCIDRs []*net.IPNet) ([]*net.IPNet, error) {
+	if allowCIDR.IP.To4() == nil {
+		return nil, fmt.Errorf("allowCIDR IP type is not v4")
+	}
+	// Remove the redundant CIDRs
+	exceptCIDRs = mergeCIDRs(exceptCIDRs)
+	newCIDRs := make([]*net.IPNet, 1)
+	newCIDRs[0] = allowCIDR
+	for _, exceptCIDR := range exceptCIDRs {
+		if exceptCIDR.IP.To4() == nil {
+			return nil, fmt.Errorf("exceptCIDR IP type is not v4")
+		}
+	beginLoop:
+		for i, indCIDR := range newCIDRs {
+			// Consider masked IP from IPNet struct
+			if indCIDR.Contains(exceptCIDR.IP.Mask(exceptCIDR.Mask)) {
+				result := diffFromCIDR(indCIDR, exceptCIDR)
+				// Delete the considered CIDR block and add resulting CIDR blocks
+				copy(newCIDRs[i:], newCIDRs[i+1:])
+				newCIDRs = newCIDRs[:len(newCIDRs)-1]
```

Comment:
I think the recommended way to remove an element from a slice is to use `append`:
```
slice = append(slice[:i], slice[i+1:]...)
```
the performance is the same as with `copy` IIRC and it's a one liner. I think it is also more readable.

Reply:
Based on the link you give, both approaches are valid.

The link does recommend to set the deleted element to 0 so that the `IPNet` object can be garbage collected immediately. @tnqn what do you think?

Reply:
My mistake, I thought a dereferenced (by slice) item will be garbage collected automatically like python does, didn't realize the item will still be referenced by the underlying array. 
It makes sense to me to set the deleted element to nil, sorry for the false comment.

Reply:
Thanks.